### PR TITLE
[DateRangeInput] Add allowSingleDayRange prop

### PR DESCRIPTION
**Pull request by cmslewis**
#### Checklist

- [x] Include tests
- [x] Update documentation

#### Changes proposed in this pull request:

- Add `allowSingleDayRange` prop to `DateRangeInput`
- Add unit tests
- Add toggle to example

#### Reviewers should focus on
 
- Tricky date comparison logic that needs to ignore hours/minutes/seconds/millis.
- Modified overlapping-dates logic.

### CODE REVIEW

File: packages/datetime/src/dateRangeInput.tsx [hunk 1]
```diff
@@ -739,10 +749,10 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
                 ? fromMomentToDate(selectedBound)
                 : undefined;
         });
-        // show only the start date if the dates overlap
-        // TODO: add different handling for the === case once
-        // allowSingleDayRange is implemented (#249)
-        return [startDate, (startDate >= endDate) ? null : endDate] as DateRange;
+        const selectedRange = this.props.allowSingleDayRange
+            ? [startDate, (clearTime(startDate) > clearTime(endDate)) ? null : endDate]
+            : [startDate, (clearTime(startDate) >= clearTime(endDate)) ? null : endDate];
```
Comment (giladgray) [inline packages/datetime/src/dateRangeInput.tsx]: what about `DateUtils.areSameDay`? that's what `DateRangePicker` [uses to implement this prop](https://github.com/palantir/blueprint/blob/master/packages/datetime/src/dateRangePicker.tsx#L442-L443).

File: packages/datetime/test/dateRangeInputTests.tsx [hunk 2]
```diff
@@ -177,6 +177,36 @@ describe("<DateRangeInput>", () => {
...
+                allowSingleDayRange={true}
+                defaultValue={[START_DATE, END_DATE]}
+            />);
+
+            // edit the end boundary
+            getEndInput(root).simulate("focus");
+
+            // deselect current end date
+            getDayElement(END_DAY).simulate("click");
+
```
Comment (giladgray) [inline packages/datetime/test/dateRangeInputTests.tsx]: such whitespace. i mean whatever, but i like keeping blocks of interaction between asserts together.

File: packages/datetime/src/dateRangeInput.tsx (hunk 1 above)
Comment (giladgray) [inline packages/datetime/src/dateRangeInput.tsx]: wait but actually... why does this get reimplemented here? can we not rely on DRP's own `allowSingleDayRange` prop implementation?

Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: Yeah there's a little duplication between the implementation for the next-selected dates (on click) and the next focused fields on click (and still more duplication/logic overlap for validation logic).

But you're right: the basic way to check this is `!DateUtils.areSameDay(startDate, endDate) && startDate > endDate`. Will edit.

**Comment by cmslewis** [comment]
@giladgray ready for re-review

File: packages/datetime/src/dateRangeInput.tsx [hunk 3]
```diff
@@ -733,16 +742,19 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
     }
 
     private getSelectedRange = () => {
-        const momentDateRange = [this.state.selectedStart, this.state.selectedEnd] as MomentDateRange;
-        const [startDate, endDate] = momentDateRange.map((selectedBound?: moment.Moment) => {
+        const { selectedStart, selectedEnd } = this.state;
+
+        // this helper function checks if the provided boundary date *would* overlap the selected
+        // other boundary date. providing the already-selected start date simply tells us if we're
+        // currently in an overlapping state.
```
Comment (leebyp) [inline packages/datetime/src/dateRangeInput.tsx]: nits: should the comment go here, or with the method? also, sentence casing

Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: we don't do sentence casing in comments AFAIK, unless it's a prop-description comment.

**Pull request merged**